lua.c

```diff
@@ -113,6 +117,85 @@ lua_apteryx_prune (lua_State *L)
     return 1;
 }
 
+bool
+lua_watch_mux (int n, const char *path, const char *value)
+{
+    bool ret;
+
+    pthread_mutex_lock (&gL_lock);
+
+    if (!gL)
+    { /* must have exclusive access to the global lua state */
+        ERROR ("LUA cb: Lua state is NULL");
+        pthread_mutex_unlock (&gL_lock);
+        return false;
+    }
+
+    lua_pushinteger (gL, lua_watch_fn_table[n]);
+    lua_gettable (gL, LUA_REGISTRYINDEX);
+    if (!lua_isfunction (gL, -1))
+    {
+        luaL_error (gL, "lua function expected");
+    }
+    lua_pushstring (gL, path);
+    lua_pushstring (gL, value);
+    lua_call (gL, 2, 1);
+    /* boolean return type is not enforced */
+    ret = !(lua_toboolean (gL, -1) == 0);
+    lua_pop (gL, 1);    /* Remove the result to clean up */
+
+    pthread_mutex_unlock (&gL_lock);
+    return ret;
+}
+
+static int
+lua_apteryx_watch (lua_State *L)
+{
+    int cb_table_index;
+
+    if ((cb_table_index = lua_cb_register (L, lua_watch_fn_table)) < 0)
+    {
+        return 0;
+    }
+
+    if (!apteryx_watch (lua_tostring (L, 1), lua_watch_cb_table[cb_table_index]))
+    {
+        ERROR ("Could not register watch at index %d\n", cb_table_index);
+        return 0;
+    }
+
+    DEBUG ("Registered watch with index %d\n", cb_table_index);
+    return 1;
+}
+
+static int
+lua_apteryx_unwatch (lua_State *L)
+{
+    int cb_table_index;
+
+    if ((cb_table_index = lua_cb_unregister (L, lua_watch_fn_table)) < 0)
+    {
+        return 0;
+    }
+
+    if (!apteryx_unwatch (lua_tostring (L, 1), lua_watch_cb_table[cb_table_index]))
+    {
+        ERROR ("Could not unregister watch at index %d\n", cb_table_index);
+    }
+
+    DEBUG ("Unregistered watch index %d\n", cb_table_index);
+    return 1;
+}
+
+int
+lua_apteryx_mainloop (lua_State *L)
+{
+    gL = L;
+    pthread_mutex_unlock (&gL_lock);
+    pause ();
+    exit (0);
```

Comment:
This should either not exit (so the script can unwatch), or go through the callback tables and forcefully unregister all active watches.